src/sbin/wsconsctl/wsconsctl.8
.\" $MirOS: src/sbin/wsconsctl/wsconsctl.8,v 1.2 2008/12/04 15:36:08 tg Exp $
.\" $OpenBSD: wsconsctl.8,v 1.18 2004/07/30 23:28:23 jmc Exp $
.\" $NetBSD: wsconsctl.8,v 1.5 1999/09/12 18:47:11 kleink Exp $
.\"
.\" Copyright (c) 1998 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Juergen Hannken-Illjes.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the NetBSD
.\"	Foundation, Inc. and its contributors.
.\" 4. Neither the name of The NetBSD Foundation nor the names of its
.\"    contributors may be used to endorse or promote products derived
.\"    from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"/
.Dd $Mdocdate: December 4 2008 $
.Dt WSCONSCTL 8
.Os
.Sh NAME
.Nm wsconsctl
.Nd get or set wscons state
.Sh SYNOPSIS
.Nm wsconsctl
.Op Fl d Ar /dev/ttyC0
.Op Fl k Ar /dev/wskdb0
.Op Fl m Ar /dev/wsmouse0
.Op Fl n
.Fl a
.Nm wsconsctl
.Op Fl d Ar /dev/ttyC0
.Op Fl k Ar /dev/wskdb0
.Op Fl m Ar /dev/wsmouse0
.Op Fl n
.Ar name ...
.Nm wsconsctl
.Op Fl d Ar /dev/ttyC0
.Op Fl k Ar /dev/wskdb0
.Op Fl m Ar /dev/wsmouse0
.Op Fl n
.Ar name Ns = Ns Ar value ...
.Nm wsconsctl
.Op Fl d Ar /dev/ttyC0
.Op Fl k Ar /dev/wskdb0
.Op Fl m Ar /dev/wsmouse0
.Op Fl n
.Ar name Ns += Ns Ar value ...
.Sh DESCRIPTION
The
.Nm
command displays or sets various wscons system driver variables.
If a list of variables is present on the command line,
.Nm
prints the current value of those variables for the specified device.
Similarily, if a list of top-level entries is present,
.Nm
prints the current values of their children.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl a
Print all device variables and their current values.
.It Fl d Ar ctldev
Set the control device for display to
.Ar ctldev .
.It Fl k Ar ctldev
Set the control device for keyboard to
.Ar ctldev .
.It Fl m Ar ctldev
Set the control device for mouse to
.Ar ctldev .
.It Fl n
Suppress printing of the variable name in the output.
.It Ar name Ns = Ns Ar value
Attempt to set the specified variable
.Ar name
to
.Ar value .
The value can be specified as either an absolute, by using the
.Ql =
symbol,
or as a relative value, by using the
.Ql +=
symbol.
See the
.Sx EXAMPLES
section for more details.
.El
.Pp
The
.Nm
utility can be used to view and modify aspects of the keyboard,
display, and mouse using the standard, machine-independent
workstation console device driver
.Xr wscons 4 .
.Pp
The keyboard type can be modified, the keyboard bell's pitch, period,
and duration can be modified,
the
.Ar typematic
value can be changed, and the keyboard encoding can be modified
to switch keys, should the user find a keyboard's default layout
difficult to use.
.Pp
There are also definitions relating to video
control and cursor control, which are not applicable to
all display types, and to text emulation and graphics
(mapped) modes.
.Pp
Use the
.Xr kbd 8
utility to determine which keyboard encodings are available for your
country.
.Sh FILES
.Bl -tag -width /usr/include/dev/wscons/wsconsio.h -compact
.It Pa /dev/ttyC0
display control device
.It Pa /dev/wskbd0
keyboard control device
.It Pa /dev/wsmouse0
mouse control device
.It Pa /etc/wsconsctl.conf
a list of parameters that get set at system startup time from
.Xr rc 8
.It Pa /usr/include/dev/wscons/wsconsio.h
keyboard/mouse/display definitions
.El
.Sh EXAMPLES
Set a UK keyboard encoding:
.Pp
.Dl # wsconsctl keyboard.encoding=uk
.Pp
Modify the current keyboard encoding so that, when the
.Ar Caps Lock
key is pressed, the same encoding sequence as
.Ar Left Control
is sent.
For a full list of keysyms, and keycodes, please refer
to the
.Ar /usr/include/dev/wscons/wsksymdef.h
file.
.Pp
.Dl # wsconsctl keyboard.map+="keysym Caps_Lock = Control_L"
.Pp
Assign the
.Ar Right Alt
key to be the group modifier (layout is changed while the key is pressed):
.Pp
.Dl # wsconsctl keyboard.map+="keycode 184=Mode_switch"
.Pp
Assign the
.Ar Right Control
key to be the lock for the group modifier.
The effect of
.Ar Mode_Lock
and
.Ar Mode_switch
is not mutually exclusive, to be consistent with
.Ar Caps Lock
and
.Ar Shift
behaviour.
.Pp
.Dl # wsconsctl keyboard.map+="keycode 157=Mode_Lock"
.Pp
Set a US keyboard encoding, with the
.Ar Caps Lock
and
.Ar Left Control
keys swapped.
The
.Ar .swapctrlcaps
encoding does not work for all national keyboard encodings.
For most purposes, the ability to set the value returned
by the
.Ar Caps Lock
key is enough \- see the previous example for details.
.Pp
.Dl # wsconsctl keyboard.encoding=us.swapctrlcaps
.Pp
Set the bell pitch to be 1200:
.Pp
.Dl # wsconsctl keyboard.bell.pitch=1200
.Pp
Add 200 to the current pitch of the bell:
.Pp
.Dl # wsconsctl keyboard.bell.pitch+=200
.Sh SEE ALSO
.Xr pckbd 4 ,
.Xr wscons 4 ,
.Xr wsconsctl.conf 5 ,
.Xr kbd 8 ,
.Xr wsconscfg 8 ,
.Xr wsfontload 8
.Sh HISTORY
The
.Nm
command first appeared in
.Nx 1.4
and
.Ox 2.8 .
The ability to print subtrees, similar to
.Xr sysctl 8 ,
and the
.Fl d , k , No and Fl m
options first appeared in
.Mx 11 .


src/sbin/wsconsctl/wsconsctl.c
/*	$OpenBSD: wsconsctl.c,v 1.16 2004/08/09 18:39:22 deraadt Exp $	*/
/*	$NetBSD: wsconsctl.c,v 1.2 1998/12/29 22:40:20 hannken Exp $ */

/*-
 * Copyright (c) 1998 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Juergen Hannken-Illjes.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the NetBSD
 *	Foundation, Inc. and its contributors.
 * 4. Neither the name of The NetBSD Foundation nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <fcntl.h>
#include <err.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include "wsconsctl.h"

__RCSID("$MirOS$");

#define PATH_KEYBOARD		"/dev/wskbd0"
#define PATH_MOUSE		"/dev/wsmouse0"
#define PATH_DISPLAY		"/dev/ttyC0"

extern const char *__progname;		/* from crt0.o */

extern struct field keyboard_field_tab[];
extern struct field mouse_field_tab[];
extern struct field display_field_tab[];

void usage(char *);

struct vartypesw {
	const char *name, *file;
	int fd;
	struct field *field_tab;
	void (*getval)(const char *pre, int);
	void (*putval)(const char *pre, int);
} typesw[] = {
	{ "keyboard", PATH_KEYBOARD, -1, keyboard_field_tab,
	  keyboard_get_values, keyboard_put_values },
	{ "mouse", PATH_MOUSE, -1, mouse_field_tab,
	  mouse_get_values, mouse_put_values },
	{ "display", PATH_DISPLAY, -1, display_field_tab,
	  display_get_values, display_put_values },
	{ NULL }
};

struct vartypesw *tab_by_name(const char *);

void
usage(char *msg)
{
	if (msg != NULL)
		fprintf(stderr, "%s: %s\n", __progname, msg);

	fprintf(stderr,
	    "usage: %s [-[kmd] ctlfile] [-n] -a\n"
	    "       %s [-[kmd] ctlfile] [-n] name ...\n"
	    "       %s [-[kmd] ctlfile] [-n] name=value ...\n"
	    "       %s [-[kmd] ctlfile] [-n] name+=value ...\n",
	    __progname, __progname, __progname, __progname);

	exit(1);
}

int
main(int argc, char *argv[])
{
	int i, ch, error = 0, aflag = 0, do_merge;
	struct vartypesw *sw = NULL;
	char *sep = "=", *p;
	struct field *f;

	while ((ch = getopt(argc, argv, "ad:k:m:nw")) != -1) {
		switch(ch) {
		case 'a':
			aflag = 1;
			break;
		case 'd':
			for (sw = typesw; sw->name; sw++)
				if (!strcmp(sw->name, "display"))
					sw->file = strdup(optarg);
			break;
		case 'k':
			for (sw = typesw; sw->name; sw++)
				if (!strcmp(sw->name, "keyboard"))
					sw->file = strdup(optarg);
			break;
		case 'm':
			for (sw = typesw; sw->name; sw++)
				if (!strcmp(sw->name, "mouse"))
					sw->file = strdup(optarg);
			break;
		case 'n':
			sep = NULL;
			break;
		case 'w':
			/* compat */
			break;
		default:
			usage(NULL);
		}
	}

	argc -= optind;
	argv += optind;

	if (argc > 0 && aflag != 0)
		usage("excess arguments after -a");

	if (aflag != 0) {
		for (sw = typesw; sw->name; sw++) {
			if (sw->fd < 0 &&
			    (sw->fd = open(sw->file, O_WRONLY)) < 0 &&
			    (sw->fd = open(sw->file, O_RDONLY)) < 0) {
				warn("%s", sw->file);
				error = 1;
				continue;
			}
			for (f = sw->field_tab; f->name; f++)
				if ((f->flags & (FLG_NOAUTO|FLG_WRONLY)) == 0)
					f->flags |= FLG_GET;
			(*sw->getval)(sw->name, sw->fd);
			for (f = sw->field_tab; f->name; f++)
				if (f->flags & FLG_DEAD)
					continue;
				else if (f->flags & FLG_NOAUTO)
					warnx("Use explicit arg to view %s.%s.",
					      sw->name, f->name);
				else if (f->flags & FLG_GET)
					pr_field(sw->name, f, sep);
		}
	} else if (argc > 0) {
		for (i = 0; i < argc; i++) {
			p = strchr(argv[i], '=');
			if (p == NULL && strchr(argv[i], '.') == NULL) {
				for (sw = typesw; sw->name; sw++)
					if (!strcmp(sw->name, argv[i]))
						break;
				if (!sw->name)
					continue;
				if (sw->fd < 0 &&
				    (sw->fd = open(sw->file, O_WRONLY)) < 0 &&
				    (sw->fd = open(sw->file, O_RDONLY)) < 0) {
					warn("%s", sw->file);
					error = 1;
					continue;
				}
				for (f = sw->field_tab; f->name; f++)
					if ((f->flags & (FLG_NOAUTO|FLG_WRONLY)) == 0)
						f->flags |= FLG_GET;
				(*sw->getval)(sw->name, sw->fd);
				for (f = sw->field_tab; f->name; f++)
					if (f->flags & FLG_DEAD)
						continue;
					else if (f->flags & FLG_NOAUTO)
						warnx("Use explicit arg to view %s.%s.",
						      sw->name, f->name);
					else if (f->flags & FLG_GET)
						pr_field(sw->name, f, sep);
				continue;
			} else if (p == NULL) {
				sw = tab_by_name(argv[i]);
				if (!sw)
					continue;
				if (sw->fd < 0 &&
				    (sw->fd = open(sw->file, O_WRONLY)) < 0 &&
				    (sw->fd = open(sw->file, O_RDONLY)) < 0) {
					warn("open: %s", sw->file);
					error = 1;
					continue;
				}
				f = field_by_name(sw->field_tab, argv[i]);
				if (f->flags & FLG_DEAD)
					continue;
				if ((f->flags & FLG_WRONLY) != 0) {
					warnx("%s: write only", argv[i]);
					continue;
				}
				f->flags |= FLG_GET;
				(*sw->getval)(sw->name, sw->fd);
				pr_field(sw->name, f, sep);
				continue;
			}
			if (p > argv[i] &&
			    (*(p - 1) == '+' || *(p - 1) == '-')) {
				do_merge = *(p - 1);
				*(p - 1) = '\0';
			} else
				do_merge = 0;
			*p++ = '\0';
			sw = tab_by_name(argv[i]);
			if (!sw)
				continue;
			if (sw->fd < 0 &&
			    (sw->fd = open(sw->file, O_WRONLY)) < 0 &&
			    (sw->fd = open(sw->file, O_RDONLY)) < 0) {
				warn("open: %s", sw->file);
				error = 1;
				continue;
			}
			f = field_by_name(sw->field_tab, argv[i]);
			if (f->flags & FLG_DEAD)
				continue;
			if ((f->flags & FLG_RDONLY) != 0) {
				warnx("%s: read only", argv[i]);
				continue;
			}
			if (do_merge || f->flags & FLG_INIT) {
				if ((f->flags & FLG_MODIFY) == 0)
					errx(1, "%s: can only be set",
					     argv[i]);
				f->flags |= FLG_GET;
				(*sw->getval)(sw->name, sw->fd);
				f->flags &= ~FLG_GET;
			}
			rd_field(f, p, do_merge);
			f->flags |= FLG_SET;
			(*sw->putval)(sw->name, sw->fd);
			f->flags &= ~FLG_SET;
		}
	} else
		usage(NULL);

	return (error);
}

struct vartypesw *
tab_by_name(const char *var)
{
	struct vartypesw *sw;
	const char *p = strchr(var, '.');

	if (!p) {
		warnx("%s: illegal variable name", var);
		return (NULL);
	}

	for (sw = typesw; sw->name; sw++)
		if (!strncmp(sw->name, var, p - var))
			break;

	if (!sw->name) {
		warnx("%s: no such variable", var);
		return (NULL);
	}

	return (sw);
}
